### PR TITLE
Close database on application shutdown

### PR DESCRIPTION
**Pull request by michplunkett**
## Addresses Issue
https://github.com/prijatelilab/PrijateliTree/pull/129#issuecomment-1875743441

## Describe Your Changes
I made the database configuration file smaller in the README, adjusted util function parameter signatures, validated that the application works in multiple views, and added `bootstrap.bundle.min.js.map` to the repository.

## Checklist Before Requesting a Review
- [x] The code runs successfully.

```commandline
2024-01-05 18:39:20 prijatelitree-web-1       | INFO:     192.168.65.1:43987 - "POST /admin/session HTTP/1.1" 302 Found
2024-01-05 18:39:20 prijatelitree-web-1       | INFO:     192.168.65.1:43987 - "GET /admin/dashboard?success=Your+session+%28ID%3A+1%29+and+first+game+%28ID%3A+2%29+have+been+created%21 HTTP/1.1" 200 OK
2024-01-05 18:41:08 prijatelitree-web-1       | INFO:     192.168.65.1:44146 - "GET /games/session/1 HTTP/1.1" 200 OK
2024-01-05 18:41:09 prijatelitree-web-1       | INFO:     192.168.65.1:44146 - "GET /games/session/1 HTTP/1.1" 200 OK
2024-01-05 18:41:12 prijatelitree-web-1       | INFO:     192.168.65.1:44146 - "GET /js/bootstrap.bundle.min.js.map HTTP/1.1" 404 Not Found
2024-01-05 18:41:12 prijatelitree-web-1       | INFO:     192.168.65.1:44146 - "GET /css/bootstrap.min.css.map HTTP/1.1" 200 OK
2024-01-05 18:41:12 prijatelitree-web-1       | INFO:     192.168.65.1:44146 - "GET /games/session/1 HTTP/1.1" 200 OK
2024-01-05 18:41:13 prijatelitree-web-1       | INFO:     192.168.65.1:44146 - "GET /css/bootstrap.min.css HTTP/1.1" 200 OK
2024-01-05 18:41:13 prijatelitree-web-1       | INFO:     192.168.65.1:44147 - "GET /css/custom.css HTTP/1.1" 200 OK
2024-01-05 18:41:13 prijatelitree-web-1       | INFO:     192.168.65.1:44148 - "GET /js/bootstrap.bundle.min.js HTTP/1.1" 200 OK
2024-01-05 18:41:13 prijatelitree-web-1       | INFO:     192.168.65.1:44150 - "GET /js/dataTables.bootstrap5.min.js HTTP/1.1" 200 OK
2024-01-05 18:41:13 prijatelitree-web-1       | INFO:     192.168.65.1:44149 - "GET /js/jquery.dataTables.min.js HTTP/1.1" 200 OK
2024-01-05 18:41:13 prijatelitree-web-1       | INFO:     192.168.65.1:44151 - "GET /js/jquery-3.6.0.min.js HTTP/1.1" 200 OK
2024-01-05 18:41:13 prijatelitree-web-1       | INFO:     192.168.65.1:44147 - "GET /images/logo.png HTTP/1.1" 200 OK
2024-01-05 18:41:13 prijatelitree-web-1       | INFO:     192.168.65.1:44147 - "GET /css/bootstrap.min.css.map HTTP/1.1" 200 OK
2024-01-05 18:41:13 prijatelitree-web-1       | INFO:     192.168.65.1:44151 - "GET /js/bootstrap.bundle.min.js.map HTTP/1.1" 404 Not Found
2024-01-05 18:43:32 prijatelitree-web-1       | INFO:     192.168.65.1:44360 - "GET /games/session/1 HTTP/1.1" 200 OK
2024-01-05 18:43:32 prijatelitree-web-1       | INFO:     192.168.65.1:44360 - "GET /css/bootstrap.min.css HTTP/1.1" 200 OK
2024-01-05 18:43:32 prijatelitree-web-1       | INFO:     192.168.65.1:44361 - "GET /css/custom.css HTTP/1.1" 200 OK
2024-01-05 18:43:32 prijatelitree-web-1       | INFO:     192.168.65.1:44363 - "GET /js/bootstrap.bundle.min.js HTTP/1.1" 200 OK
2024-01-05 18:43:32 prijatelitree-web-1       | INFO:     192.168.65.1:44362 - "GET /js/jquery-3.6.0.min.js HTTP/1.1" 200 OK
2024-01-05 18:43:32 prijatelitree-web-1       | INFO:     192.168.65.1:44364 - "GET /js/jquery.dataTables.min.js HTTP/1.1" 200 OK
2024-01-05 18:43:32 prijatelitree-web-1       | INFO:     192.168.65.1:44365 - "GET /js/dataTables.bootstrap5.min.js HTTP/1.1" 200 OK
2024-01-05 18:43:32 prijatelitree-web-1       | INFO:     192.168.65.1:44361 - "GET /images/logo.png HTTP/1.1" 200 OK
2024-01-05 18:43:32 prijatelitree-web-1       | INFO:     192.168.65.1:44361 - "GET /js/bootstrap.bundle.min.js.map HTTP/1.1" 404 Not Found
2024-01-05 18:43:32 prijatelitree-web-1       | INFO:     192.168.65.1:44361 - "GET /css/bootstrap.min.css.map HTTP/1.1" 200 OK
2024-01-05 18:44:10 prijatelitree-web-1       | INFO:     192.168.65.1:44422 - "GET /games/session/1 HTTP/1.1" 200 OK
2024-01-05 18:45:20 prijatelitree-web-1       | INFO:     192.168.65.1:44523 - "GET /games/1/player/1/ready HTTP/1.1" 200 OK
2024-01-05 18:45:26 prijatelitree-web-1       | INFO:     192.168.65.1:44538 - "GET /games/1/player/1/ready HTTP/1.1" 200 OK
2024-01-05 18:45:31 prijatelitree-web-1       | INFO:     192.168.65.1:44541 - "GET /games/1/player/1/survey HTTP/1.1" 200 OK
...
2024-01-05 18:51:01 prijatelitree-web-1       | INFO:     192.168.65.1:44877 - "GET / HTTP/1.1" 200 OK
2024-01-05 18:51:12 prijatelitree-web-1       | INFO:     192.168.65.1:44885 - "GET /admin/login HTTP/1.1" 200 OK
2024-01-05 18:51:28 prijatelitree-web-1       | INFO:     192.168.65.1:44900 - "GET /js/bootstrap.bundle.min.js.map HTTP/1.1" 200 OK
2024-01-05 18:51:28 prijatelitree-web-1       | INFO:     192.168.65.1:44901 - "GET /css/bootstrap.min.css.map HTTP/1.1" 200 OK
2024-01-05 18:51:29 prijatelitree-web-1       | INFO:     192.168.65.1:44901 - "GET /admin/login HTTP/1.1" 200 OK
2024-01-05 18:51:29 prijatelitree-web-1       | INFO:     192.168.65.1:44901 - "GET /css/bootstrap.min.css HTTP/1.1" 200 OK
2024-01-05 18:51:29 prijatelitree-web-1       | INFO:     192.168.65.1:44900 - "GET /css/custom.css HTTP/1.1" 200 OK
2024-01-05 18:51:29 prijatelitree-web-1       | INFO:     192.168.65.1:44902 - "GET /js/bootstrap.bundle.min.js HTTP/1.1" 200 OK
2024-01-05 18:51:29 prijatelitree-web-1       | INFO:     192.168.65.1:44905 - "GET /js/jquery-3.6.0.min.js HTTP/1.1" 200 OK
2024-01-05 18:51:29 prijatelitree-web-1       | INFO:     192.168.65.1:44903 - "GET /js/jquery.dataTables.min.js HTTP/1.1" 200 OK
2024-01-05 18:51:29 prijatelitree-web-1       | INFO:     192.168.65.1:44904 - "GET /js/dataTables.bootstrap5.min.js HTTP/1.1" 200 OK
2024-01-05 18:51:29 prijatelitree-web-1       | INFO:     192.168.65.1:44900 - "GET /images/logo.png HTTP/1.1" 200 OK
2024-01-05 18:51:29 prijatelitree-web-1       | INFO:     192.168.65.1:44900 - "GET /css/bootstrap.min.css.map HTTP/1.1" 200 OK
2024-01-05 18:51:29 prijatelitree-web-1       | INFO:     192.168.65.1:44905 - "GET /js/bootstrap.bundle.min.js.map HTTP/1.1" 200 OK
```

Screenshot 2024-01-05 at 6 45 17 PM
Screenshot 2024-01-05 at 6 45 25 PM
Screenshot 2024-01-05 at 6 45 30 PM
Screenshot 2024-01-05 at 6 45 37 PM

### CODE REVIEW

File: prijateli_tree/app/utils/games.py [hunk 1]
```diff
@@ -87,7 +83,7 @@ def get_bag_color(bag) -> str:
     return correct_answer
 
 
-def get_current_round(game_id: int, db: Session = Depends(Database)) -> int:
+def get_current_round(game_id: int, db: Session) -> int:
```
Comment (michplunkett) [inline prijateli_tree/app/utils/games.py:86]: These are not FastAPI specific functions, we don't need the `Depends`.

File: prijateli_tree/app/main.py [hunk 2]
```diff
@@ -43,3 +44,9 @@
 @app.get("/", response_class=HTMLResponse)
 def home(request: Request) -> Response:
     return templates.TemplateResponse("home_page.html", {"request": request})
+
+
+@app.on_event("shutdown")
+def close_external_connections():
+    # Close database connection on application shutdown
+    Database.close_connection()
```
Comment (michplunkett) [inline prijateli_tree/app/main.py:52]:
```commandline
2024-01-05 19:06:43 prijatelitree-web-1       | INFO:     192.168.65.1:46073 - "GET / HTTP/1.1" 200 OK
2024-01-05 19:06:43 prijatelitree-web-1       | INFO:     192.168.65.1:46073 - "GET /css/bootstrap.min.css HTTP/1.1" 304 Not Modified
2024-01-05 19:06:43 prijatelitree-web-1       | INFO:     192.168.65.1:46075 - "GET /css/custom.css HTTP/1.1" 304 Not Modified
2024-01-05 19:06:43 prijatelitree-web-1       | INFO:     192.168.65.1:46074 - "GET /js/bootstrap.bundle.min.js HTTP/1.1" 304 Not Modified
2024-01-05 19:06:43 prijatelitree-web-1       | INFO:     192.168.65.1:46077 - "GET /js/jquery.dataTables.min.js HTTP/1.1" 304 Not Modified
2024-01-05 19:06:43 prijatelitree-web-1       | INFO:     192.168.65.1:46073 - "GET /images/logo.png HTTP/1.1" 304 Not Modified
2024-01-05 19:06:43 prijatelitree-web-1       | INFO:     192.168.65.1:46078 - "GET /js/dataTables.bootstrap5.min.js HTTP/1.1" 304 Not Modified
2024-01-05 19:06:43 prijatelitree-web-1       | INFO:     192.168.65.1:46076 - "GET /js/jquery-3.6.0.min.js HTTP/1.1" 304 Not Modified
2024-01-05 19:06:51 prijatelitree-web-1       | INFO:     Shutting down
2024-01-05 19:06:51 prijatelitree-web-1       | INFO:     Waiting for application shutdown.
2024-01-05 19:06:51 prijatelitree-web-1       | INFO:   Database connection closed at 2024-01-06 01:06:51.937285
2024-01-05 19:06:51 prijatelitree-web-1       | INFO:     Application shutdown complete.
2024-01-05 19:06:51 prijatelitree-web-1       | INFO:     Finished server process [8]
2024-01-05 19:06:52 prijatelitree-web-1       | INFO:     Stopping reloader process [1]
prijatelitree-web-1 exited with code 0
prijatelitree-web-1 exited with code 0
prijatelitree-postgres-1  | 2024-01-06T01:06:52.156633509Z 2024-01-06 01:06:52.156 UTC [1] LOG:  received fast shutdown request
2024-01-05 19:06:52 prijatelitree-postgres-1  | 2024-01-06 01:06:52.158 UTC [1] LOG:  aborting any active transactions
2024-01-05 19:06:52 prijatelitree-postgres-1  | 2024-01-06 01:06:52.159 UTC [1] LOG:  background worker "logical replication launcher" (PID 32) exited with exit code 1
2024-01-05 19:06:52 prijatelitree-postgres-1  | 2024-01-06 01:06:52.160 UTC [27] LOG:  shutting down
2024-01-05 19:06:52 prijatelitree-postgres-1  | 2024-01-06 01:06:52.160 UTC [27] LOG:  checkpoint starting: shutdown immediate
2024-01-05 19:06:52 prijatelitree-postgres-1  | 2024-01-06 01:06:52.163 UTC [27] LOG:  checkpoint complete: wrote 3 buffers (0.0%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.001 s, sync=0.001 s, total=0.004 s; sync files=2, longest=0.001 s, average=0.001 s; distance=0 kB, estimate=0 kB; lsn=0/1A027C0, redo lsn=0/1A027C0
2024-01-05 19:06:52 prijatelitree-postgres-1  | 2024-01-06 01:06:52.165 UTC [1] LOG:  database system is shut down
prijatelitree-postgres-1 exited with code 0
```

File: prijateli_tree/app/main.py [hunk 3]
```diff
@@ -43,3 +44,9 @@
 @app.get("/", response_class=HTMLResponse)
 def home(request: Request) -> Response:
     return templates.TemplateResponse("home_page.html", {"request": request})
+
+
+@app.on_event("shutdown")
```
Comment (michplunkett) [inline prijateli_tree/app/main.py:49]: https://fastapi.tiangolo.com/advanced/events/#shutdown-event

File: prijateli_tree/app/main.py (hunk 2 above)
Comment (anisfeld) [reply prijateli_tree/app/main.py:52]: When does a shutdown event occur?

**Review by anisfeld — APPROVED** [review]
Looks good and it works. Some details are a bit hazy to me, but that's okay. I asked a question in the review about shutdown though that I'd like to know more about.

File: prijateli_tree/app/main.py (hunk 2 above)
Comment (michplunkett) [reply prijateli_tree/app/main.py:52]: The shutdown event occurs when the FastAPI application closes. I simulated that using the `make stop` command. For more information you can look here: https://fastapi.tiangolo.com/advanced/events/#shutdown-event

**Pull request merged**